File: 328-odd-even-linked-list/328-odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head==None or head.next==None: return head
        oddHead = None
        oddTail = None
        evenHead = None
        evenTail = None
        current = head
        i = 1
        
        while current!=None:
            if i%2==1:
                if oddHead == None:
                    oddHead = current
                    oddTail = current
                else: 
                    oddTail.next = current
                    oddTail = oddTail.next
            else:
                if evenHead == None:
                    evenHead = current
                    evenTail = current
                else:
                    evenTail.next=current
                    evenTail = evenTail.next
            
            current = current.next
            i+=1
        
        evenTail.next= None
        oddTail.next = evenHead
        
        return oddHead
                
                
        
        

File: 328-odd-even-linked-list/328-odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if(head ==null || head.next==null) return head;
        ListNode oddHead = null;
        ListNode oddTail = null;
        ListNode evenHead = null;
        ListNode evenTail = null;
        ListNode curNode = head;
        int pointer = 1;
        
        while(curNode != null){
            if(pointer%2==1){
                if(oddHead == null){
                    oddHead = curNode;
                    oddTail = curNode;
                } else {
                    oddTail.next = curNode;
                    oddTail = oddTail.next;
                }
            } else{
                if(evenHead == null){
                    evenHead = curNode;
                    evenTail = curNode;
                } else{
                    evenTail.next = curNode;
                    evenTail = evenTail.next;
                }
            }
            
            curNode = curNode.next;
            pointer++;
        }
        
        evenTail.next = null;
        oddTail.next = evenHead;
        
        return oddHead;
        
    }
}